src/btree/tuple_chunk.c

```diff
@@ -0,0 +1,521 @@
+				   Page page, void *key, BTreeKeyType keyType,
+				   OffsetNumber *itemOffset)
+{
+	bool		isLeaf = O_PAGE_IS(page, LEAF),
+				nextKey;
+	BTreeKeyType midKeyType;
+	OffsetNumber low,
+				mid,
+				high;
+	int			targetCmpVal,
+				result;
+
+	if (partial && !load_chunk_partially(desc, partial, page))
+		return false;
+
+	if (desc->chunkItemsCount == 0)
+	{
+		*itemOffset = 0;
+		return true;
+	}
+
+	midKeyType = isLeaf ? BTreeKeyLeafTuple : BTreeKeyNonLeafKey;
+
+	low = 0;
+	high = desc->chunkItemsCount - 1;
+	nextKey = (!isLeaf && keyType != BTreeKeyPageHiKey);
+
+	/* Shouldn't look for hikey on leafs, because we're already here */
+	Assert(!(isLeaf && keyType == BTreeKeyPageHiKey));
+
+	/*
+	 * Binary search to find the first key on the page >= `key`, or first page
+	 * key > `key` when nextkey is true.
+	 *
+	 * For nextkey=false (cmp=1), the loop invariant is: all slots before
+	 * `low` are < `key`, all slots at or after `high` are >= `key`.
+	 *
+	 * For nextkey=true (cmp=0), the loop invariant is: all slots before `low`
+	 * are <= `key`, all slots at or after `high` are > `key`.
+	 *
+	 * We can fall out when `high` == `low`.
+	 */
+	high++;						/* establish the loop invariant for high */
+
+	targetCmpVal = nextKey ? 0 : 1; /* a target value of cmpFunc() */
+
+	/*
+	 * Don't pass BTreeHiKey to comparison function, we've set nextkey flag
+	 * instead.
+	 */
+	if (keyType == BTreeKeyPageHiKey)
+		keyType = BTreeKeyNonLeafKey;
+
+	while (high > low)
+	{
+		mid = low + ((high - low) / 2);
+
+		if (!isLeaf && mid == 0 && *itemOffset == 0)
+			result = 1;
+		else
+		{
+			OTuple		midTuple;
+
+			*itemOffset = mid;
+			midTuple = read_tuple_chunk_tuple(desc, page, *itemOffset);
+			result = desc->ops->keycmp(desc->treeDesc, key, keyType,
+									   &midTuple, midKeyType);
+		}
+
+		if (result >= targetCmpVal)
+			low = mid + 1;
+		else
+			high = mid;
+	}
+
+	*itemOffset = low;
+
+	return true;
+}
+
+bool
+tuple_chunk_read_tuple(BTreeChunkDesc *desc, PartialPageState *partial,
+					   Page page, OffsetNumber itemOffset,
+					   Pointer *tupleHeader, OTuple *tuple)
+{
+	if (partial && !load_chunk_partially(desc, partial, page))
+		return false;
+
+	Assert(itemOffset < desc->chunkItemsCount);
+
+	if (O_PAGE_IS(page, LEAF))
+		read_tuple_chunk_leaf_item(desc, page, itemOffset,
+								   (BTreeLeafTuphdr **) tupleHeader, tuple);
+	else
+		read_tuple_chunk_internal_item(desc, page, itemOffset,
+									   (BTreeNonLeafTuphdr **) tupleHeader,
+									   tuple);
+
+	return true;
+}
+
+void
+tuple_chunk_builder_init(BTreeChunkBuilderData *builderData,
+						 BTreeDescr *treeDesc,
+						 BTreeChunkType chunkType, Size size)
+{
+	builderData->treeDesc = treeDesc;
+	builderData->builderData = palloc(size);
+	builderData->builderMaxSize = size;
+	builderData->builderSize = 0;
+	builderData->chunkType = chunkType;
+}
+
+int32
+tuple_chunk_builder_estimate(BTreeChunkBuilderData *builderData,
```

Comment:
Also, need to take into account the size of new items[] element.

Reply:
Fixed the function `tuple_chunk_builder_estimate`.